### PR TITLE
Reduce canvas memory footprint for better iOS stability

### LINKED ISSUE #9578: Image layer renderer uses more memory than necessary
For image sources that load data from e.g. a WMS with a pixel ratio of 1, we create a canvas with the size for the display pixel ratio instead of the image ratio of the source. So for HiDPI devices, the resulting canvas is twice as big as it needs to be, using 4 times the memory.

### LINKED ISSUE #12908: Crashes on iOS 15
When using polygons filled with a canvas pattern, WebView crashes after a few moves of the map.

If we do not set map.pixelRatio: 1 then crashes occur even if the polygons are not filled with a pattern.

On Chromium 91 (macOS 11.6, Window 11 arm) and Android (tested on 5.1 and 11) WebView, crashes do not occur (even if map.pixelRatio: 3)

---
iPhone 12 Pro Max
iOS: 15.0.2
WebView: Mozilla/5.0 (iPhone; CPU iPhone OS 14_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148

An example of an image we use to fill a polygon:

Err Detail:

[Process] 0x10c5decd8 - GPUProcessProxy::didClose:
[Process] 0x10c5decd8 - GPUProcessProxy::gpuProcessExited: reason=crash
[Process] 0x10aecc900 - [PID=8847] WebProcessProxy::gpuProcessExited: reason=0
[Process] 0x104e0e0b8 - WebProcessPool::gpuProcessDidExit: GPU Process has crashed more than 2 times in the last 30 seconds, terminating all WebProcesses

### LINKED ISSUE #13585: iphone11ProMax ios15.4.1  Error rendering MVT
**TypeError: null is not an object (evaluating 'context.scale') /common.js:133248:14 draw_**
**Uncaught (in promise)  Object {name: "TypeError",message: "null is not an object (evaluating 'ctx.fillStyle = 'rgba(0,0,0,0)'')"**
**stack: "getBlankImage@http://172.20.10.2:8080/common.js:54943:6\nhandleImageError_**
**TypeError: null is not an object (evaluating 'context.scale') /common.js**
**TypeError: null is not an object (evaluating 'context.canvas') **

### PR DESCRIPTION
**Pull request by ahocevar**
The goal of this pull request is to reduce the canvas memory footprint, which should solve several issues in Safari, especially on iOS.

Each commit is self-contained and addresses one specific problem.

The biggest change is the way we handle opacity in the canvas renderers. Previously, we set layer opacity on the canvas element. This means that for each layer that had an opacity configured, we needed a separate canvas, and could not share rendering contexts. Now we handle opacity using `globalAlpha` on the canvas. For tile layers, we sacrifice tile transitions when a layer opacity is set. For vector layers with opacity, we now use a short-lived composition canvas.

Another important change is the canvas size for tile and image layers with a tile or image pixel ratio lower or higher than the device pixel ratio. Previously, the target canvas had the same resolution as the input data. Now it has the resolution matching the device pixel ratio.

The changed opacity and pixel ratio handling combined now allow us to almost always share rendering contexts, except when combining Canvas, WebGL and custom renderers, and of course when layers have different `className`s set.

Other changes include the reuse and release of canvases as described in https://pqina.nl/blog/total-canvas-memory-use-exceeds-the-maximum-limit/ (also see #13321), and smaller canvases on rotated views, where we now only render tiles that are actually visible in the viewport.

These changes not only make crashes on iOS due to hitting canvas size limits or excessive canvas rendering less like, they also improve performance - especially on rotated views.

To see the improved stability on iOS, browse to https://deploy-preview-13823--ol-site.netlify.app/examples/mapbox-style.html with an iPhone in landscape view, and compare with https://openlayers.org/en/v6.14.1/examples/mapbox-style.html. The latter will crash almost immediately when zooming and panning on a rotated view. The former will only crash after a long time zooming and panning really, really fast and repeatedly.

Fixes #12908
Fixes #13585 
Fixes #9578

### CODE REVIEW

File: src/ol/renderer/canvas/TileLayer.js [hunk 1]
```diff
@@ -711,17 +727,35 @@ class CanvasTileLayerRenderer extends CanvasLayerRenderer {
...
         for (y = tileRange.minY; y <= tileRange.maxY; ++y) {
+          const tileVisible =
+            !rotation ||
+            intersectsLinearRing(
+              viewport,
+              0,
+              viewport.length,
+              2,
+              tileGrid.getTileCoordExtent([z, x, y])
+            );
```
Comment (MoonE) [inline src/ol/renderer/canvas/TileLayer.js:754]: Why not `continue` here when the tile is not visible?

File: src/ol/renderer/canvas/TileLayer.js [hunk 2]
```diff
@@ -735,7 +769,7 @@ class CanvasTileLayerRenderer extends CanvasLayerRenderer {
             if (opt_tileCallback !== undefined) {
               opt_tileCallback(tile);
             }
-          } else {
+          } else if (tileVisible) {
             tileSource.useTile(z, x, y, projection);
```
Comment (MoonE) [inline src/ol/renderer/canvas/TileLayer.js:773]: This is unrelated to the changes in this PR, but shouldn't the used tiles increase the `tileCount` when they are cached?

**Comment by wayofthefuture** [comment]
Looking forward to trying this! Thanks!

File: src/ol/renderer/canvas/TileLayer.js (hunk 1 above)
Comment (ahocevar) [reply src/ol/renderer/canvas/TileLayer.js]: Thanks for your sharp eye, @MoonE. I updated the commit with your suggested improvement.

**Comment by mike-000** [comment]
Reprojection can set the `globalCompositeOperation` to `lighter` and `imageSmoothingEnabled` false.  If canvas is to returned to a pool these should be restored to their default values.

**Comment by ahocevar** [comment]
> Reprojection can set the `globalCompositeOperation` to `lighter` and `imageSmoothingEnabled` false. If canvas is to returned to a pool these should be restored to their default values.

These will be reset by `releaseCanvas()`, which sets width and height and hence clears all settings of the context.

**Review by jahow — APPROVED** [review]
I've tested many examples and everything seems to work just fine (except the visual glitch I mentioned on a separate comment). The changes make sense as well. I used an emulator for testing on iOS devices and sadly could not reproduce the crash there (probably because I wasn't in landscape and wasn't panning quick enough), so if someone could confirm that everything works fine on a separate iOS device it would be great!

The visual glitch is probably due to the rounding logic when computing the desired canvas size in https://github.com/openlayers/openlayers/pull/13823/commits/6ffbfa1a4af81a68a02ee13c8328203b5ffe13e3#diff-d63a4fd4c80f2dafc266d6c1a32961c8adca7a4777e1b94d7642542b930aae19, so no big deal most likely.

Overall quite an impressive optimization, thanks!

**Comment by ahocevar** [comment]
Thanks for the review, @jahow. I fixed the image layer rotation regression, and an old image layer extent clipping issue, which can be seen when looking at the westernmost part of Washington state in the wms-image example, which was sometimes clipped during rotation in main, and now always completely rendered.

**Comment by tschaub** [comment]
It looks like this may have introduced an issue on [NASA Worldview](https://worldview.earthdata.nasa.gov/).  See #14171 and https://github.com/nasa-gibs/worldview/pull/3996.

**Comment by zboros1** [comment]
This problem still persists!

IOS devices continue to crash from OpenLayers versions 6.10 - 7.1.
Customer notices this when I upgraded from 6.5 to 6.15.1 and they could not use the application in any IOS device.  We do have several layers loading (about 10), publishing roughly 80 layers from a map, which was not the issue with previous versions of Open Layers on IOS 15.  Users also noticed slower  map loading and performance on desktop web browsers when upgrading to 6.15 and later.

I rolled back to OpenLayers 6.9, which fixed the crash on IOS devices but panning stopped after a few zoom and pan operations still.

I eventually had to go to version 6.6.1 in which everything is working fine.  No crashes, no delays. Note, I originally started in version 6.5.  So earlier versions of OpenLayers seem to work fine with IOS 15.7 but not the later versions.  I guess we'll stay here until this problem is resolved.

I use TileLayer with ImageWMS source.  I tried converting them to WebGLTile but there was some funky image mirroring on  zoom, so abandoned that idea.

**Comment by tschaub** [comment]
@zboros1 - If you are able to put together an example that demonstrates the problem, that would make it easier to track down.

**Comment by ahocevar** [comment]
@zboros1 

> I use TileLayer with ImageWMS source

You should be using a TileWMS source with a TileLayer. But like @tschaub says, seeing an example that demonstrates the problem would make it easier to identify the problem.

**Comment by zboros1** [comment]
Oops, I stated that incorrectly, sorry.  We use TileWMS with TileLayer.  We also use ImageWMS with ImageLayer.  Here is the snippet for the TileWMS:
 
        //Setting layer styles
       if (styles && styles.length > 0) {
          params = {'LAYERS': layerdefs, 'STYLES': styles, 'TILED': true};
        } else {
            params = { 'LAYERS': layerdefs, 'TILED': true};
        }

      //Setting Layer Source
       var imgSource = new TileWMS({
          url: url,
          params: params,
          hidpi: false,  //keep hidpi false so tiles are kept at 256,256 for webcache
          serverType: server_type,
          //crossOrigin: 'anonymous',
          tileLoadFunction: customImageLoader,
          transition: 0,
        });

        //Creating layer
        layerWMS = new TileLayer({
            source: imgSource,
            className: layer_name, //className required for selecting layers
        });

      //Setting zoom dependency
      if(maxRes > 0) {
              layerWMS.setMaxResolution(maxRes);
      }
      if (minRes > 0) {
              layerWMS.setMinResolution(minRes);
      }

    //Setting some values used for the layer later
    layerWMS.set('layerid', layer_id);
    layerWMS.set('name', layer_name);
    layerWMS.set('visible', display_on);
    layerWMS.set('selectable', selectable);
    layerWMS.set('fields', fields);
    layerWMS.set('GeometryField', fieldGeom);
    layerWMS.set('DefaultField', fieldSelect);

    //Set display order
    layerWMS.setZIndex(zIndex);
   
    //Add layer to map
    map.addLayer(layerWMS);

**Comment by mike-000** [comment]
Similar problem to #14171  You have `name` to identify the layer, you do not need `className`.

**Comment by zboros1** [comment]
So it seems so.  I removed the className and it seems to be working ok in version 6.15.1.  I will do further testing to make sure, but so far seem to not crash on mobile now.
Thanks

**Pull request merged**